3.8.0_to_4.0.0_mysql.sql

```diff
@@ -352,6 +352,69 @@ DROP TABLE `COMPONENTMETRICS`;
 DROP TABLE `CPEREFERENCE`;
 
 
+-- Find dangling components that no project depends on anymore and remove them
+
+DELIMITER $$
+
+CREATE PROCEDURE cleanup_components()
+BEGIN
+    DECLARE v_component_id BIGINT(20);
+    DECLARE v_done BIT DEFAULT FALSE;
+
+    DECLARE component_cursor CURSOR FOR
+        SELECT c.`ID`
+        FROM `COMPONENT` c
+        WHERE (SELECT COUNT(*) FROM `DEPENDENCY` d WHERE c.`ID` = d.`COMPONENT_ID`) = 0;
```

Comment:
@nscuro this subquery-based solution is not the most efficient approach because the subquery gets executed for each row in the `COMPONENT` table.

I would like to recommend using a `LEFT JOIN` between the `COMPONENT` and `DEPENDENCY` tables and having a null check in the `WHERE` condition. This will yield a significantly faster cursor. It would look something like this:
```
SELECT c.`ID`
FROM `COMPONENT` c
LEFT JOIN `DEPENDENCY` d ON c.`ID` = d.`COMPONENT_ID`
WHERE d.`COMPONENT_ID` IS NULL;
```